### PR TITLE
Add new `manual_unwrap_or_default` lint

### PR DESCRIPTION
**Pull request by GuillaumeGomez**
This adds a new lint checking if a `match` or a `if let` can be replaced with `unwrap_or_default`.

----

changelog: Add new [`manual_unwrap_or_default`] lint

### CODE REVIEW

**Comment by GuillaumeGomez** [comment]
Also I'm not a big fan of the lint name but couldn't find anything better. If anyone has a suggestion, it'd be very appreciated. :)

**Comment by llogiq** [comment]
How about `manual_unwrap_or_default`?

File: clippy_lints/src/match_option_and_default.rs [hunk 1]
```diff
@@ -0,0 +1,183 @@
...
+        && implements_trait(cx, match_ty, default_trait_id, &[])
+        // We now get the bodies for both the `Some` and `None` arms.
+        && let Some(((body_some, binding_id), body_none)) = get_some_and_none_bodies(cx, arm1, arm2)
+        // We check that the `Some(x) => x` doesn't do anything apart "returning" the value in `Some`.
+        && let ExprKind::Path(QPath::Resolved(_, path)) = body_some.peel_blocks().kind
+        && let Res::Local(local_id) = path.res
+        && local_id == binding_id
+        // We now check the `None` arm is calling a method equivalent to `Default::default`.
+        && let body_none = body_none.peel_blocks()
+        && let ExprKind::Call(_, &[]) = body_none.kind
```
Comment (llogiq) [inline clippy_lints/src/match_option_and_default.rs:126]: Is that really needed here, when in the next line we check for defaultness?

Comment (GuillaumeGomez) [reply clippy_lints/src/match_option_and_default.rs:126]: Just checked the `is_default_equivalent` implementation and you're absolutely right! Even worse: this check can actually be pretty bad as it excludes some potential cases I don't cover from what I saw. Good catch!

**Comment by GuillaumeGomez** [comment]
Renamed and applied the suggestion. Thanks for both! :)

**Pull request merged**